daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -483,41 +481,82 @@ private[validation] object Typing {
       AlphaEquiv.alphaEquiv(t1, t2) ||
         AlphaEquiv.alphaEquiv(expandTypeSynonyms(t1), expandTypeSynonyms(t2))
 
-    private def checkIfaceImplementations(
+    private def checkGenImplementation(
         tplTcon: TypeConName,
-        impls: Map[TypeConName, TemplateImplements],
+        ifaceTcon: TypeConName,
+        implMethods: List[(MethodName, Expr)],
     ): Unit = {
+      val DefInterfaceSignature(requires, _, _, methods, _, _) =
+        handleLookup(ctx, interface.lookupInterface(ifaceTcon))
 
-      impls.foreach { case (iface, impl) =>
-        val DefInterfaceSignature(requires, _, _, methods, _, _) =
-          handleLookup(ctx, interface.lookupInterface(impl.interfaceId))
-
-        requires
-          .filterNot(impls.contains)
-          .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, iface, required))
+      requires
+        .filterNot(required =>
+          interface.lookupInterfaceCoImplements(tplTcon, required).isRight ||
+            interface.lookupTemplateImplements(tplTcon, required).isRight
+        )
+        .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, ifaceTcon, required))
 
-        methods.values.foreach { (method: InterfaceMethod) =>
-          if (!impl.methods.contains(method.name))
-            throw EMissingInterfaceMethod(ctx, tplTcon, impl.interfaceId, method.name)
-        }
-        impl.methods.values.foreach { (tplMethod: TemplateImplementsMethod) =>
-          methods.get(tplMethod.name) match {
-            case None =>
-              throw EUnknownInterfaceMethod(ctx, tplTcon, impl.interfaceId, tplMethod.name)
-            case Some(method) =>
-              checkExpr(tplMethod.value, method.returnType)
-          }
+      methods.values.foreach { (method: InterfaceMethod) =>
+        if (!implMethods.exists { case (name, _) => name == method.name })
+          throw EMissingInterfaceMethod(ctx, tplTcon, ifaceTcon, method.name)
```

Comment:
Some of these error messages are written from the perspective of the template. I would like to add a parameter `: ImplementsOrCoImplements` or similar to choose the right wording, which the two callsites of `checkGenImplementation` would set to the appropriate value

Reply:
sounds very sensible

Reply:
Sounds good to me.

Reply:
I'll do this in a future PR